File: ledger/cbor.go
package ledger

import (
	"encoding/binary"
	"fmt"
	"sort"
)

// EncodeCanonical produces a deterministic CBOR encoding of v per RFC 8949's
// Core Deterministic Encoding Requirements (§4.2.1): every integer uses its
// shortest valid form, arrays and maps use definite lengths, and map keys
// are sorted by the bytewise lexicographic order of their own encoded
// bytes — not the original string, and not length-first (that was the
// older RFC 7049 canonical rule; RFC 8949 changed it).
//
// Supported types: nil, bool, int64, uint64, string, []byte, []any,
// map[string]any. Floats are deliberately unsupported. docs/ARCHITECTURE.md
// §2.2 requires fixed-point encoding wherever a value would otherwise need
// a fraction, precisely so this encoder never has to pick an IEEE-754
// canonicalization rule (NaN payloads, -0 vs 0, etc. all have more than one
// "correct" byte representation — the property this package exists to
// avoid).
func EncodeCanonical(v any) ([]byte, error) {
	return encodeValue(nil, v)
}

func encodeValue(buf []byte, v any) ([]byte, error) {
	switch t := v.(type) {
	case nil:
		return append(buf, 0xf6), nil // simple value 22 (null)
	case bool:
		if t {
			return append(buf, 0xf5), nil // simple value 21 (true)
		}
		return append(buf, 0xf4), nil // simple value 20 (false)
	case int:
		return encodeInt(buf, int64(t)), nil
	case int64:
		return encodeInt(buf, t), nil
	case uint64:
		return encodeHead(buf, 0, t), nil
	case string:
		return encodeBytesLike(buf, 3, []byte(t)), nil
	case []byte:
		return encodeBytesLike(buf, 2, t), nil
	case []any:
		buf = encodeHead(buf, 4, uint64(len(t)))
		var err error
		for _, item := range t {
			buf, err = encodeValue(buf, item)
			if err != nil {
				return nil, err
			}
		}
		return buf, nil
	case map[string]any:
		return encodeMap(buf, t)
	default:
		return nil, fmt.Errorf("ledger: canonical CBOR encoder does not support type %T (add fixed-point/typed handling explicitly — do not silently coerce)", v)
	}
}

func encodeInt(buf []byte, n int64) []byte {
	if n >= 0 {
		return encodeHead(buf, 0, uint64(n))
	}
	// CBOR major type 1 (negative integer) encodes the argument as
	// -(n+1); e.g. -1 encodes argument 0, -256 encodes argument 255.
	// n+1 cannot overflow int64 even at n == math.MinInt64: the result
	// (math.MinInt64+1) negates cleanly to math.MaxInt64.
	return encodeHead(buf, 1, uint64(-(n + 1)))
}

func encodeBytesLike(buf []byte, major byte, b []byte) []byte {
	buf = encodeHead(buf, major, uint64(len(b)))
	return append(buf, b...)
}

// encodeHead writes a CBOR major-type/argument head using the shortest
// valid form, as Core Deterministic Encoding requires (no non-shortest-form
// integers, no indefinite-length markers).
func encodeHead(buf []byte, major byte, n uint64) []byte {
	m := major << 5
	switch {
	case n < 24:
		return append(buf, m|byte(n))
	case n <= 0xff:
		return append(buf, m|24, byte(n))
	case n <= 0xffff:
		var b [2]byte
		binary.BigEndian.PutUint16(b[:], uint16(n))
		return append(append(buf, m|25), b[:]...)
	case n <= 0xffffffff:
		var b [4]byte
		binary.BigEndian.PutUint32(b[:], uint32(n))
		return append(append(buf, m|26), b[:]...)
	default:
		var b [8]byte
		binary.BigEndian.PutUint64(b[:], n)
		return append(append(buf, m|27), b[:]...)
	}
}

func encodeMap(buf []byte, m map[string]any) ([]byte, error) {
	type entry struct {
		encodedKey []byte
		value      any
	}
	entries := make([]entry, 0, len(m))
	for k, v := range m {
		entries = append(entries, entry{encodedKey: encodeBytesLike(nil, 3, []byte(k)), value: v})
	}
	// RFC 8949 §4.2.1 rule 3: sort by the bytewise lexicographic order of
	// each key's own deterministic encoding. This is what makes the
	// output independent of Go's (deliberately randomized) map iteration
	// order — see cbor_test.go's determinism property test.
	sort.Slice(entries, func(i, j int) bool {
		return bytesLess(entries[i].encodedKey, entries[j].encodedKey)
	})

	buf = encodeHead(buf, 5, uint64(len(entries)))
	var err error
	for _, e := range entries {
		buf = append(buf, e.encodedKey...)
		buf, err = encodeValue(buf, e.value)
		if err != nil {
			return nil, err
		}
	}
	return buf, nil
}

func bytesLess(a, b []byte) bool {
	n := len(a)
	if len(b) < n {
		n = len(b)
	}
	for i := 0; i < n; i++ {
		if a[i] != b[i] {
			return a[i] < b[i]
		}
	}
	return len(a) < len(b)
}

File: ledger/verify/src/cbor.rs
//! Deterministic CBOR encoding — RFC 8949 Core Deterministic Encoding
//! Requirements (§4.2.1).
//!
//! This is an INDEPENDENT re-implementation of `ledger/cbor.go`'s exact
//! algorithm, not a shared library the two languages both call into.
//! Agreement between this and the Go implementation — proven by the
//! golden-vector test in `tests/golden.rs` — is what actually demonstrates
//! cross-language determinism. A shared library would only prove that one
//! implementation is self-consistent.

/// The canonical value shapes this encoder accepts. Deliberately no float
/// variant — docs/ARCHITECTURE.md §2.2 requires fixed-point wherever a
/// value would otherwise need a fraction, so this type can't even
/// represent one.
#[derive(Debug, Clone, PartialEq)]
pub enum Value {
    Null,
    Bool(bool),
    Int(i64),
    UInt(u64),
    Text(String),
    Bytes(Vec<u8>),
    Array(Vec<Value>),
    /// Insertion order does not matter — `encode_map` sorts by each key's
    /// own encoded bytes before emitting, exactly like ledger/cbor.go's
    /// encodeMap. A `BTreeMap<String, Value>` would sort by the RAW string
    /// instead, which is the wrong order (see cbor_test.go's
    /// TestEncodeCanonical_KeyOrderIsBytewiseOnEncodedKey) — that's why
    /// this is a plain Vec of pairs, not a BTreeMap.
    Map(Vec<(String, Value)>),
}

pub fn encode_canonical(v: &Value) -> Vec<u8> {
    let mut buf = Vec::new();
    encode_value(&mut buf, v);
    buf
}

fn encode_value(buf: &mut Vec<u8>, v: &Value) {
    match v {
        Value::Null => buf.push(0xf6),
        Value::Bool(true) => buf.push(0xf5),
        Value::Bool(false) => buf.push(0xf4),
        Value::Int(n) => encode_int(buf, *n),
        Value::UInt(n) => encode_head(buf, 0, *n),
        Value::Text(s) => encode_bytes_like(buf, 3, s.as_bytes()),
        Value::Bytes(b) => encode_bytes_like(buf, 2, b),
        Value::Array(items) => {
            encode_head(buf, 4, items.len() as u64);
            for item in items {
                encode_value(buf, item);
            }
        }
        Value::Map(entries) => encode_map(buf, entries),
    }
}

fn encode_int(buf: &mut Vec<u8>, n: i64) {
    if n >= 0 {
        encode_head(buf, 0, n as u64);
    } else {
        // CBOR major type 1 argument = -(n+1). Computed in i128 so this
        // cannot overflow even at n == i64::MIN (mirrors the comment in
        // ledger/cbor.go's encodeInt, and TestEncodeCanonical_NegativeMinInt64DoesNotOverflow).
        let arg = (-(n as i128) - 1) as u64;
        encode_head(buf, 1, arg);
    }
}

fn encode_bytes_like(buf: &mut Vec<u8>, major: u8, b: &[u8]) {
    encode_head(buf, major, b.len() as u64);
    buf.extend_from_slice(b);
}

/// Writes a CBOR major-type/argument head using the shortest valid form —
/// Core Deterministic Encoding forbids non-shortest-form integers and
/// indefinite-length markers.
fn encode_head(buf: &mut Vec<u8>, major: u8, n: u64) {
    let m = major << 5;
    if n < 24 {
        buf.push(m | (n as u8));
    } else if n <= 0xff {
        buf.push(m | 24);
        buf.push(n as u8);
    } else if n <= 0xffff {
        buf.push(m | 25);
        buf.extend_from_slice(&(n as u16).to_be_bytes());
    } else if n <= 0xffff_ffff {
        buf.push(m | 26);
        buf.extend_from_slice(&(n as u32).to_be_bytes());
    } else {
        buf.push(m | 27);
        buf.extend_from_slice(&n.to_be_bytes());
    }
}

fn encode_map(buf: &mut Vec<u8>, entries: &[(String, Value)]) {
    let mut encoded: Vec<(Vec<u8>, &Value)> = entries
        .iter()
        .map(|(k, v)| {
            let mut kbuf = Vec::new();
            encode_bytes_like(&mut kbuf, 3, k.as_bytes());
            (kbuf, v)
        })
        .collect();
    // RFC 8949 §4.2.1 rule 3: sort by the bytewise lexicographic order of
    // each key's own deterministic encoding. Vec<u8>'s Ord is exactly
    // bytewise lexicographic comparison, so a plain sort_by on the encoded
    // key bytes is correct here — matches ledger/cbor.go's bytesLess.
    encoded.sort_by(|a, b| a.0.cmp(&b.0));

    encode_head(buf, 5, encoded.len() as u64);
    for (kbytes, v) in encoded {
        buf.extend_from_slice(&kbytes);
        encode_value(buf, v);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // Same ground-truth vectors as ledger/cbor_test.go's
    // TestEncodeCanonical_KnownVectors — hand-derived from RFC 8949 §3,
    // not from either implementation.
    #[test]
    fn known_vectors() {
        assert_eq!(encode_canonical(&Value::Int(0)), vec![0x00]);
        assert_eq!(encode_canonical(&Value::Int(23)), vec![0x17]);
        assert_eq!(encode_canonical(&Value::Int(24)), vec![0x18, 0x18]);
        assert_eq!(encode_canonical(&Value::Int(255)), vec![0x18, 0xff]);
        assert_eq!(encode_canonical(&Value::Int(256)), vec![0x19, 0x01, 0x00]);
        assert_eq!(encode_canonical(&Value::Int(-1)), vec![0x20]);
        assert_eq!(encode_canonical(&Value::Int(-24)), vec![0x37]);
        assert_eq!(encode_canonical(&Value::Int(-25)), vec![0x38, 0x18]);
        assert_eq!(encode_canonical(&Value::Text("".into())), vec![0x60]);
        assert_eq!(encode_canonical(&Value::Text("a".into())), vec![0x61, 0x61]);
        assert_eq!(encode_canonical(&Value::Array(vec![])), vec![0x80]);
        assert_eq!(
            encode_canonical(&Value::Array(vec![Value::Int(1), Value::Int(2), Value::Int(3)])),
            vec![0x83, 0x01, 0x02, 0x03]
        );
        assert_eq!(encode_canonical(&Value::Map(vec![])), vec![0xa0]);
        assert_eq!(encode_canonical(&Value::Bool(true)), vec![0xf5]);
        assert_eq!(encode_canonical(&Value::Bool(false)), vec![0xf4]);
        assert_eq!(encode_canonical(&Value::Null), vec![0xf6]);
        assert_eq!(
            encode_canonical(&Value::Bytes(vec![0xde, 0xad])),
            vec![0x42, 0xde, 0xad]
        );
    }

    #[test]
    fn map_sorted_by_encoded_key_not_raw_string() {
        // "b" (len 1, head 0x61) sorts before "a0" (len 2, head 0x62),
        // even though raw-string order would put "a0" first. See the
        // matching Go test for the full explanation.
        let got = encode_canonical(&Value::Map(vec![
            ("a0".into(), Value::Int(2)),
            ("b".into(), Value::Int(1)),
        ]));
        let want = vec![
            0xa2, 0x61, 0x62, 0x01, // "b": 1
            0x62, 0x61, 0x30, 0x02, // "a0": 2
        ];
        assert_eq!(got, want);
    }

    #[test]
    fn min_i64_does_not_overflow() {
        let got = encode_canonical(&Value::Int(i64::MIN));
        let want = vec![0x3b, 0x7f, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff];
        assert_eq!(got, want);
    }
}
